apiAuditLogs/apiAuditLogs.go

```diff
@@ -0,0 +1,290 @@
+		"getUsersByIds":                        true,
+		"getUsersByNames":                      true,
+		"getUserStatus":                        true,
+		"getUserStatusesByIds":                 true,
+		"getUserTermsOfService":                true,
+		"getWebappPlugins":                     true,
+		"listAutocompleteCommands":             true,
+		"listCommands":                         true,
+		"logout":                               true,
+		"openDialog":                           true,
+		"patchChannelModerations":              true,
+		"pinPost":                              true,
+		"postLog":                              true,
+		"pushNotificationAck":                  true,
+		"saveReaction":                         true,
+		"searchAllChannels":                    true,
+		"searchArchivedChannelsForTeam":        true,
+		"searchChannelsForTeam":                true,
+		"searchEmojis":                         true,
+		"searchGroupChannels":                  true,
+		"searchPosts":                          true,
+		"searchTeams":                          true,
+		"searchUserAccessTokens":               true,
+		"searchUsers":                          true,
+		"setPostUnread":                        true,
+		"submitDialog":                         true,
+		"teamExists":                           true,
+		"teamMembersMinusGroupMembers":         true,
+		"testElasticsearch":                    true,
+		"testEmail":                            true,
+		"testLdap":                             true,
+		"testS3":                               true,
+		"testSiteURL":                          true,
+		"unpinPost":                            true,
+		"updateUserStatus":                     true,
+		"uploadFileStream":                     true,
+		"viewChannel":                          true,
+	}
+
+	if pass.Pkg.Path() != "github.com/mattermost/mattermost-server/v5/api4" {
+		return nil, nil
+	}
+
+	inspect := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	nodeFilter := []ast.Node{
+		(*ast.FuncDecl)(nil),
+	}
+	inspect.Preorder(nodeFilter, func(n ast.Node) {
+		funDecl := n.(*ast.FuncDecl)
+
+		if strings.HasSuffix(pass.Fset.File(n.Pos()).Name(), "_test.go") {
+			return
+		}
+		if strings.HasSuffix(pass.Fset.File(n.Pos()).Name(), "apitestlib.go") {
+			return
+		}
+		if obj, ok := pass.TypesInfo.Defs[funDecl.Name].(*types.Func); ok {
+			var fact facts.IsApiHandler
+			if !pass.ImportObjectFact(obj, &fact) {
+				return
+			}
+		}
+		if whiteList[funDecl.Name.Name] {
+			return
+		}
+		initializationFound := false
+		logCallFound := false
+		successCallFound := false
+		ast.Inspect(funDecl, func(node ast.Node) bool {
+			switch n := node.(type) {
+			case *ast.CallExpr:
+				fun, ok := n.Fun.(*ast.SelectorExpr)
+				if !ok {
+					return true
+				}
+
+				ident, ok := fun.X.(*ast.Ident)
+				if !ok {
+					return true
+				}
+
+				// must have a auditRec.Success()
+				if ident.Name == "auditRec" && fun.Sel.Name == "Success" {
```

Comment:
It's too bad that the true type of the Ident is not available here.  A rule like "must have identifier of type `audit.Record` that calls `Success`" would be slightly less brittle.  Not big deal, just means the identifier must always be named "auditRec".

Reply:
We can improve it checking the variable name created during the `MakeAuditRecord` initialization.

Reply:
We can indeed get the type of `Ident`. You can see here: https://github.com/agnivade/sqlargs/blob/master/sqlargs.go#L139-L145.

Reply:
Done! Thanks for the pointer @agnivade, very useful.